### PR TITLE
Add claimable functionality

### PR DESCRIPTION
**Pull request by ECWireless**

### CODE REVIEW

File: components/Modals/CreateItemModal.tsx [hunk 1]
```diff
@@ -218,18 +219,48 @@ export const CreateItemModal: React.FC<CreateItemModalProps> = ({
...
+            ]);
+            claimable = tree.root as `0x${string}`;
+
+            const jsonTree = JSON.stringify(tree.dump());
+            const data = {
+              itemId: game.items.length,
+              gameAddress: game.id,
+              tree: jsonTree,
+            };
+            const res = await fetch('/api/setTree', {
```
Comment (dan13ram) [inline components/Modals/CreateItemModal.tsx:247]: Need to add authentication to this route.

File: components/Modals/CreateItemModal.tsx [hunk 2]
```diff
@@ -218,18 +221,55 @@ export const CreateItemModal: React.FC<CreateItemModalProps> = ({
...
+
+          if (trimmedAddresses.length === 0) {
+            claimable = pad('0x01');
+          } else {
+            const leaves: ClaimableItemLeaf[] = trimmedAddresses.map(
+              address => {
+                return [
+                  BigInt(game.items.length),
+                  getAddress(address),
+                  BigInt(itemSupply),
```
Comment (dan13ram) [inline components/Modals/CreateItemModal.tsx:238]: We need to update this to set only the amount we want the user to be able to claim.

File: hooks/useClaimableTree.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  const { data, error, mutate, isLoading, isValidating } = useSWR<
+    StandardMerkleTree<ClaimableItemLeaf> | null,
+    Error,
+    FetcherInput
+  >(input, fetcher, {
+    isPaused: () => !gameAddress || !itemId || gameAddress === zeroAddress,
+  });
+
+  return {
+    loading: isLoading || isValidating || data === undefined,
```
Comment (ECWireless) [inline hooks/useClaimableTree.ts:55]: For some reason `loading` is hanging for a long time when the `ClaimItemModal` first pops up. Seems like it has to do with `data` not getting fetched immediately

**Review by ECWireless — COMMENTED** [review]
@dan13ram everything is working for me! The only issue I'm seeing is that `ClaimItemModal` has "loading..." hanging for a strangely long time. Seems like maybe `useSWR` is getting paused until some other state in the modal changes

File: components/Modals/CreateItemModal.tsx [hunk 4]
```diff
@@ -88,14 +99,15 @@ export const CreateItemModal: React.FC<CreateItemModalProps> = ({
...
-    const trimmedAddresses = addresses.map(address => address.trim());
-    return (
-      trimmedAddresses.some(address => !isAddress(address)) &&
-      !!whitelistedClaimers
+  const invalidClaimableAddressList = useMemo(() => {
+    return claimableAddressList.some(
+      ({ address, amount }) =>
+        !isAddress(address) ||
+        BigInt(amount) <= BigInt(0) ||
+        BigInt(amount) > maxUint256 ||
```
[the remark this answers is not in the document]
Comment (dan13ram) [reply components/Modals/CreateItemModal.tsx:107]: Yes 100%.. But actually it has to be the sum of all amounts.

File: components/Modals/CreateItemModal.tsx [hunk 5]
```diff
@@ -218,18 +230,63 @@ export const CreateItemModal: React.FC<CreateItemModalProps> = ({
...
-
-        // if (claimableToggle) {
-        //   const addresses = whitelistedClaimers.split(',');
-        //   const trimmedAddresses = addresses.map(address => address.trim());
-
-        //   if (trimmedAddresses.length === 0) {
-        //     claimable = pad('0x01');
-        //   }
-        // }
+        let claimable = pad('0x00');
```
[the remark this answers is not in the document]
Comment (dan13ram) [reply components/Modals/CreateItemModal.tsx:233]: Yes you're right

File: components/Modals/CreateItemModal.tsx [hunk 6]
```diff
@@ -218,18 +230,63 @@ export const CreateItemModal: React.FC<CreateItemModalProps> = ({
...
+            claimable = tree.root as `0x${string}`;
+
+            const jsonTree = JSON.stringify(tree.dump());
+            const data = {
+              itemId: game.items.length,
+              gameAddress: game.id,
+              tree: jsonTree,
+            };
+
+            const signature = await walletClient.signMessage({
```
[the remark this answers is not in the document]
Comment (dan13ram) [reply components/Modals/CreateItemModal.tsx:263]: Oh yes. We may need to handle that better. Do you have any ideas ?

File: hooks/useClaimableTree.ts (hunk 3 above)
Comment (dan13ram) [reply hooks/useClaimableTree.ts:55]: Yeah. I have seen this. I dont know why it hangs so long. Do you have any idea ?

Comment (ECWireless) [reply hooks/useClaimableTree.ts:55]: Going to try to debug it today

File: components/Modals/CreateItemModal.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (ECWireless) [reply components/Modals/CreateItemModal.tsx:263]: @dan13ram maybe for now (for the RaidGuild game setup at least) we can just add one of us as a GameMaster to more quickly set up the game (and leave the RG DAO as the owner)?

File: components/Modals/CreateItemModal.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (ECWireless) [reply components/Modals/CreateItemModal.tsx:113]: Fixed in 5d458940d8d69f235d605ae00400dfc0a86d833c

File: components/Modals/CreateItemModal.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (ECWireless) [reply components/Modals/CreateItemModal.tsx]: Fixed in 5d458940d8d69f235d605ae00400dfc0a86d833c

**Pull request merged**